### PR TITLE
MueLu: Adjust vector processing in blocked smoother Apply() method

### PR DESCRIPTION
**Pull request by maxfirmbach**
@trilinos/muelu

## Motivation

If we enter the `BlockedSmoothers` with a reordered/nested matrix, things go wrong in the `Apply()` method if it is called with plain `MultiVector` objects.
It is essential to first do a blocking with a plain `BlockedMap` and afterwards do a reordering based on the `ReorderManager` if the operator is nested. Currently the first step can't be done in a correct way if a nested map is applied, due to a reordered operator.

Thus changing the initial processing of the vectors in the `Apply()` method of each `BlockedSmoother`.

This might not be the perfect way to do this, but it worked for me quite well.

## Stakeholder Feedback

Related to physics-based block-preconditioning.

## Additional Information

Also interesting for @mayrmt and @Max-pert

### CODE REVIEW

File: packages/muelu/src/Smoothers/BlockedSmoothers/MueLu_BlockedGaussSeidelSmoother_def.hpp [hunk 1]
```diff
@@ -201,85 +201,73 @@ namespace MueLu {
...
+    RCP<Xpetra::ReorderedBlockedCrsMatrix<Scalar,LocalOrdinal,GlobalOrdinal,Node> > rbA =
+            Teuchos::rcp_dynamic_cast<Xpetra::ReorderedBlockedCrsMatrix<Scalar,LocalOrdinal,GlobalOrdinal,Node> >(bA);
+
+    if(rbA.is_null() == false) {
+      // A is a ReorderedBlockedCrsMatrix and thus uses nested maps, retrieve BlockedCrsMatrix and use plain blocked
+      // map for the construction of the blocked multivectors
+
+      // check type of X vector
+      if (bX.is_null() == true) {
+        RCP<MultiVector> test = Teuchos::rcp(new BlockedMultiVector(rbA->getBlockedCrsMatrix()->getBlockedDomainMap(), rcpX));
```
Comment (mayrmt) [inline packages/muelu/src/Smoothers/BlockedSmoothers/MueLu_BlockedGaussSeidelSmoother_def.hpp:221]: Maybe a more descriptive variable name?
```suggestion
        RCP<MultiVector> vectorWithBlockedMap = Teuchos::rcp(new BlockedMultiVector(rbA->getBlockedCrsMatrix()->getBlockedDomainMap(), rcpX));
```

File: packages/muelu/src/Smoothers/BlockedSmoothers/MueLu_BlockedGaussSeidelSmoother_def.hpp [hunk 2]
```diff
@@ -201,85 +201,73 @@ namespace MueLu {
...
+        rcpX.swap(test);
+        bCopyResultX = true;
+        bReorderX = true;
+      }
+
+      // check type of B vector
+      if (bB.is_null() == true) {
+        RCP<const MultiVector> test = Teuchos::rcp(new BlockedMultiVector(rbA->getBlockedCrsMatrix()->getBlockedRangeMap(), rcpB));
```
Comment (mayrmt) [inline packages/muelu/src/Smoothers/BlockedSmoothers/MueLu_BlockedGaussSeidelSmoother_def.hpp:229]: Maybe a more descriptive variable name?
```suggestion
        RCP<const MultiVector> vectorWithBlockedMap = Teuchos::rcp(new BlockedMultiVector(rbA->getBlockedCrsMatrix()->getBlockedRangeMap(), rcpB));
```

File: packages/muelu/src/Smoothers/BlockedSmoothers/MueLu_BlockedGaussSeidelSmoother_def.hpp [hunk 3]
```diff
@@ -201,85 +201,73 @@ namespace MueLu {
...
+        rcpB.swap(test);
+        bReorderB = true;
+      }
     }
+    else {
+      // A is a BlockedCrsMatrix and uses a plain blocked map
+      if (bX.is_null() == true) {
+        RCP<MultiVector> test = Teuchos::rcp(new BlockedMultiVector(bA->getBlockedDomainMap(), rcpX));
```
Comment (mayrmt) [inline packages/muelu/src/Smoothers/BlockedSmoothers/MueLu_BlockedGaussSeidelSmoother_def.hpp:237]: Maybe a more descriptive variable name?
```suggestion
        RCP<MultiVector> vectorWithBlockedMap = Teuchos::rcp(new BlockedMultiVector(bA->getBlockedDomainMap(), rcpX));
```

File: packages/muelu/src/Smoothers/BlockedSmoothers/MueLu_BlockedGaussSeidelSmoother_def.hpp [hunk 4]
```diff
@@ -201,85 +201,73 @@ namespace MueLu {
...
+        rcpX.swap(test);
+        bCopyResultX = true;
+      }
 
-    if(bB.is_null() == true) {
-      RCP<const MultiVector> test = Teuchos::rcp(new BlockedMultiVector(bA->getBlockedRangeMap(),rcpB));
-      rcpB.swap(test);
+      if(bB.is_null() == true) {
+        RCP<const MultiVector> test = Teuchos::rcp(new BlockedMultiVector(bA->getBlockedRangeMap(),rcpB));
```
Comment (mayrmt) [inline packages/muelu/src/Smoothers/BlockedSmoothers/MueLu_BlockedGaussSeidelSmoother_def.hpp:243]: Maybe a more descriptive variable name?
```suggestion
        RCP<const MultiVector> vectorWithBlockedMap = Teuchos::rcp(new BlockedMultiVector(bA->getBlockedRangeMap(),rcpB));
```

File: packages/muelu/src/Smoothers/BlockedSmoothers/MueLu_BlockedGaussSeidelSmoother_def.hpp [hunk 5]
```diff
@@ -201,85 +201,73 @@ namespace MueLu {
...
+
       Teuchos::RCP<const Xpetra::BlockReorderManager > brm = rbA->getBlockReorderManager();
 
       // check type of X vector
-      if(bX->getBlockedMap()->getNumMaps() != bA->getDomainMapExtractor()->NumMaps()) {
+      if(bX->getBlockedMap()->getNumMaps() != bA->getDomainMapExtractor()->NumMaps() || bReorderX) {
         // X is a blocked multi vector but incompatible to the reordered blocked operator A
-        Teuchos::RCP<MultiVector> test =
-            buildReorderedBlockedMultiVector(brm, bX);
+        Teuchos::RCP<MultiVector> test = buildReorderedBlockedMultiVector(brm, bX);
```
Comment (mayrmt) [inline packages/muelu/src/Smoothers/BlockedSmoothers/MueLu_BlockedGaussSeidelSmoother_def.hpp:260]: Maybe a more descriptive variable name?
```suggestion
        Teuchos::RCP<MultiVector> reorderedBlockedVector = buildReorderedBlockedMultiVector(brm, bX);
```

File: packages/muelu/src/Smoothers/BlockedSmoothers/MueLu_BlockedGaussSeidelSmoother_def.hpp [hunk 6]
```diff
@@ -201,85 +201,73 @@ namespace MueLu {
...
         rcpX.swap(test);
       }
-      if(bB->getBlockedMap()->getNumMaps() != bA->getRangeMapExtractor()->NumMaps()) {
+
+      if(bB->getBlockedMap()->getNumMaps() != bA->getRangeMapExtractor()->NumMaps() || bReorderB) {
         // B is a blocked multi vector but incompatible to the reordered blocked operator A
-        Teuchos::RCP<const MultiVector> test =
-            buildReorderedBlockedMultiVector(brm, bB);
+        Teuchos::RCP<const MultiVector> test = buildReorderedBlockedMultiVector(brm, bB);
```
Comment (mayrmt) [inline packages/muelu/src/Smoothers/BlockedSmoothers/MueLu_BlockedGaussSeidelSmoother_def.hpp:266]: Maybe a more descriptive variable name?
```suggestion
        Teuchos::RCP<const MultiVector> reorderedBlockedVector = buildReorderedBlockedMultiVector(brm, bB);
```

**Review by mayrmt — CHANGES_REQUESTED** [review]
@maxfirmbach The idea looks good to me. I left some comments for one of the block smoothers, but they actually apply to the other smoothers as well.

I understand, that these changes enable new use cases. Can you please make sure to cover them by tests, so that we don't accidentally loose these new capabilities in the future?

**Comment by maxfirmbach** [comment]
I added unit testing routines for the block smoothers using a plain, un-nested multivector for X and RHS with a reordered blocked crs matrix A.

**Comment by maxfirmbach** [comment]
Updated the naming of temporary variables as proposed by @mayrmt. This is again ready for testing :smiley:.

**Review by mayrmt — APPROVED** [review]
Thanks, @maxfirmbach! LGTM.

**Comment by mayrmt** [comment]
@maxfirmbach Ready to be merged?

**Comment by maxfirmbach** [comment]
@mayrmt Yes 👍

**Pull request merged**